core/src/server/method_response.rs

```diff
@@ -0,0 +1,537 @@
+#[derive(Debug, Clone, Default)]
+pub struct BatchResponseBuilder {
+	/// Serialized JSON-RPC response,
+	result: String,
+	/// Max limit for the batch
+	max_response_size: usize,
+}
+
+impl BatchResponseBuilder {
+	/// Create a new batch response builder with limit.
+	pub fn new_with_limit(limit: usize) -> Self {
+		let mut initial = String::with_capacity(2048);
+		initial.push('[');
+
+		Self { result: initial, max_response_size: limit }
+	}
+
+	/// Append a result from an individual method to the batch response.
+	///
+	/// Fails if the max limit is exceeded and returns to error response to
+	/// return early in order to not process method call responses which are thrown away anyway.
+	pub fn append(&mut self, response: &MethodResponse) -> Result<(), MethodResponse> {
+		// `,` will occupy one extra byte for each entry
+		// on the last item the `,` is replaced by `]`.
+		let len = response.result.len() + self.result.len() + 1;
+
+		if len > self.max_response_size {
+			Err(MethodResponse::error(Id::Null, reject_too_big_batch_response(self.max_response_size)))
+		} else {
+			self.result.push_str(&response.result);
+			self.result.push(',');
+			Ok(())
+		}
+	}
+
+	/// Check if the batch is empty.
+	pub fn is_empty(&self) -> bool {
+		self.result.len() <= 1
+	}
+
+	/// Finish the batch response
+	pub fn finish(mut self) -> BatchResponse {
+		if self.result.len() == 1 {
+			BatchResponse(batch_response_error(Id::Null, ErrorObject::from(ErrorCode::InvalidRequest)))
+		} else {
+			self.result.pop();
+			self.result.push(']');
+			BatchResponse(self.result)
+		}
+	}
+}
+
+/// Serialized batch response.
+#[derive(Debug, Clone)]
+pub struct BatchResponse(String);
+
+/// Create a JSON-RPC error response.
+pub fn batch_response_error(id: Id, err: impl Into<ErrorObject<'static>>) -> String {
+	let err = InnerResponsePayload::error_borrowed(err);
+	serde_json::to_string(&Response::new(err, id)).expect("JSON serialization infallible; qed")
+}
+
+/// Similar to [`jsonrpsee_types::ResponsePayload`] but possible to with an async-like
+/// API to detect when a method response has been sent.
+#[derive(Debug)]
+pub struct ResponsePayload<'a, T>
+where
+	T: Clone,
+{
+	inner: InnerResponsePayload<'a, T>,
+	on_exit: Option<NotifyKind>,
+}
+
+impl<'a, T: Clone> From<InnerResponsePayload<'a, T>> for ResponsePayload<'a, T> {
+	fn from(inner: InnerResponsePayload<'a, T>) -> Self {
+		Self { inner, on_exit: None }
+	}
+}
+
+impl<'a, T> ResponsePayload<'a, T>
+where
+	T: Clone,
+{
+	/// Create successful an owned response payload.
+	pub fn result(t: T) -> Self {
+		InnerResponsePayload::Result(StdCow::Owned(t)).into()
+	}
+
+	/// Create successful borrowed response payload.
+	pub fn result_borrowed(t: &'a T) -> Self {
+		InnerResponsePayload::Result(StdCow::Borrowed(t)).into()
+	}
+
+	/// Get a notification when the successful method response has been sent out.
+	pub fn notify_on_success(mut self, tx: MethodResponseNotifyTx) -> Self {
```

Comment:
Another small thought/question about this API; is there any benefit to the user passing in a `MethodResponseNotifyTx`?

You could have `response.notify_on_success(&mut self) -> MethodResponseNotifyRx` for instance, to save having to import the relevant type and create a notify channel pair.

Sample usage might be:

```rust
let res = ResponsePayload::result(..);

let on_complete = res.notify_on_completion();

tokio::spawn(async move {
    if let false = on_complete.await {
        return
    };

    // do things after message sent and successful.
});

res
```

Reply:
Good idea :)